From 9c4bf9040562c2f17285bace53e8f6c0d3f9f7e0 Mon Sep 17 00:00:00 2001
From: Vishal Sagar <vishal.sagar@xilinx.com>
Date: Fri, 10 Jan 2020 17:24:56 -0800
Subject: [PATCH 0803/1851] v4l: xilinx: xcsi2rxss: Add v4.1 compatible string

commit ab7a893fe1d759efa05eb7f85c1d3e45895c3e7b from
https://github.com/Xilinx/linux-xlnx.git

This patch adds the v4.1 compatible string to make the driver compatible
with IP v4.1 where the IIC has been removed from the subsystem. Now if
the DPHY is present then its register offset will be 0x10000 only as per
the Sub-core Address Offsets table in PG232. The iic-present property
shouldn't be present when the compatible string is v4.1.

Signed-off-by: Vishal Sagar <vishal.sagar@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/media/platform/xilinx/xilinx-csi2rxss.c | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/drivers/media/platform/xilinx/xilinx-csi2rxss.c b/drivers/media/platform/xilinx/xilinx-csi2rxss.c
index 4368fbcc1319..b10ee088f014 100644
--- a/drivers/media/platform/xilinx/xilinx-csi2rxss.c
+++ b/drivers/media/platform/xilinx/xilinx-csi2rxss.c
@@ -332,6 +332,7 @@
 #define XCSI_GET_BITSET_STR(val, mask)	(val) & (mask) ? "true" : "false"
 
 #define XCSI_CLK_PROP		BIT(0)
+#define XCSI_DPHY_PROP		BIT(1)
 
 /**
  * struct xcsi2rxss_feature - dt or IP property structure
@@ -512,6 +513,10 @@ struct xcsi2rxss_state {
 	bool suspended;
 };
 
+static const struct xcsi2rxss_feature xlnx_csi2rxss_v4_1 = {
+	.flags = XCSI_CLK_PROP | XCSI_DPHY_PROP,
+};
+
 static const struct xcsi2rxss_feature xlnx_csi2rxss_v4_0 = {
 	.flags = XCSI_CLK_PROP,
 };
@@ -527,6 +532,8 @@ static const struct of_device_id xcsi2rxss_of_id_table[] = {
 		.data = &xlnx_csi2rxss_v2_0 },
 	{ .compatible = "xlnx,mipi-csi2-rx-subsystem-4.0",
 		.data = &xlnx_csi2rxss_v4_0 },
+	{ .compatible = "xlnx,mipi-csi2-rx-subsystem-4.1",
+		.data = &xlnx_csi2rxss_v4_1 },
 	{ }
 };
 MODULE_DEVICE_TABLE(of, xcsi2rxss_of_id_table);
@@ -1623,6 +1630,15 @@ static int xcsi2rxss_parse_of(struct xcsi2rxss_state *xcsi2rxss)
 	dev_dbg(core->dev, "IIC present property = %s\n",
 			iic_present ? "Present" : "Absent");
 
+	if (iic_present && (core->cfg->flags & XCSI_DPHY_PROP)) {
+		/*
+		 * In IP v4.1 the DPHY offset is 0x10000, if present,
+		 * and the iic is removed from subsystem.
+		 */
+		dev_err(core->dev, "Invalid case - IIC present!");
+		return -EINVAL;
+	}
+
 	if (core->dphy_present) {
 		if (iic_present)
 			core->dphy_offset = 0x20000;
-- 
2.31.1

